### ISSUE TITLE
`box` event doesn't trigger on edges. (FeatureRequest)

### ISSUE THREAD

**Issue by umdstu** [cytoscape/cytoscape.js#2728]
**Issue type**
Feature request

**Description of new feature**
I noticed that the Collections `box` event only fires on nodes, not edges. This limitation isn't mentioned in the docs, so it may also be a bug. I'm unsure if the fact that calling `select()` on an edge does not actually appear to toggle it's selected state could be related to this or not.

**Motivation for new feature**
 I would like to be able to drag to select multiple edges like one does for nodes. This would be useful for bulk changing or deleting duplicate edges between two nodes.

Thanks as always.

**Comment by maxkfranz** [comment]
Could you show an example, perhaps a video, of your edge issue?

**Comment by umdstu** [comment]
Sure, I modified an existing demo to show this: https://codepen.io/umdstu/pen/QWyPPgg

If you drag over a node, it gets selected. If you drag over the to edges between Jerry and Elaine, the edges do not get selected.
What you do get is a `boxstart` and a `boxend` event, but no `boxselect` event. Boxselect is only fired if the box selection area contains a node. Furthermore, `boxselect` includes the start and end x,y coordinates as well as the width and height inside of the events target.bodyBounds object, whereas `boxstart` and `boxend` naturally do not include this (only position respectively).

What I would expect to happen is when dragging over the two edges in the above demo, that they become selected just like they do when you drag and select both of the nodes. Alternatively, if the boxselect event fired with the same bounding box even if no nodes were in the bounds (perhaps this would essentially be the implementation to achieve the former).

Hope that helps explain, thanks

**Comment by umdstu** [comment]
Update: I figured out what is happening.
The mouseup event is firing last, after the select() completes, which I believe is expected. However, inside of that event handler, because selectionType() is not 'additive' and multSelKeyDown is not down*, anything selected gets unselected (`cy.$(isSelected).unmerge(box).select();`).

*This works fine in my sample demo because in that version the pan/zoom functionality is not inverted. When you invert the pan/zoom functionality (in my case drag is the default behavior, pan/zoom requires the modifier keys), the keys aren't ever down when dragging.

Because of how the library is coded, it seems the unexpected side effect of inverting the behavior is that multiSelectKeyDown won't ever be accurate.

I didn't check the side effects, but I modified the mouseup event to instead check if `!cy.boxSelectionEnabled()` rather than `!multSelKeyDown` and that seems to work while maintaining the intended behavior, but I didn't think about this too deeply.

Thoughts?

**Comment by maxkfranz** [comment]
There may be other actions that are dependent on modifier keys.  One that comes to mind is additive selection by tapping (e.g. shift+tap).  Is that action affected by your change?

**Comment by umdstu** [comment]
I don't believe so based on this line down to line 982: https://github.com/cytoscape/cytoscape.js/blob/unstable/src/extensions/renderer/base/load-listeners.js#L964

But there are other places there `multSelKeyDown` is used to determine whether to perform actions that I feel like aren't working for me (unknowingly until now) since i manually control pan/drag/zoom. I'd need to dig into it more to see what I've lost, but I think maybe if instead of depending on `multSelKeyDown` you depend on whether the functions that users can control are enabled or disabled (eg. `cy.panningEnabled`) in these locations the library would function properly while being more flexible.

**Comment by umdstu** [comment]
I think my last comment (and the one before that) probably should have been added to #2729, not this one.

This feature request is unrelated to the multSelKeyDown issues discussed above. 

The bottom line is that drag to select edges doesn't select edges, and i think it would be useful to support that. I was able to accomplish it by listening to boxstart and boxend to capture positions, then creating a polygon out of the bounds of the dragged area. I then iterate through all of the edges and seeing if the beginning or end position of each segment, or any part of the segment itself, intersected with that polygon, and if so, toggle .select() on the edge.

I apologize for mixing these two issues up.
